128-139/128-139.md

```diff
@@ -1,9 +1,485 @@
+main.main.func1()
+
+/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go :65 +0x47
+Previous write at 0x000001228340 by goroutine 7: main.main.func1()
+/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go
+
+:75 +0x68
+Goroutine 8 (running) created at: main.main()
+/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go :62 +0xab
+Goroutine 7 (finished) created at: main.main()
+/Users/hoanhan/work/hoanhan101/ultimate-go/go/concurrency/data_race_1.go :62 +0xab
+==================
+Final Counter: 4
+Found 1 data race(s)
+exit status 66
+```
+
+##### Atomic Functions
+
+```go
+package main
+
+import (
+    "fmt"
+    "runtime"
+    "sync"
+    "sync/atomic"
+)
+```
+
+counter is a variable incremented by all Goroutine. Notice that it's not just an int but int64. We are being very specific about the precision because the atomic function requires us to do so.
+
+```go
+var counter int64
+
+func main() {
+```
+
+Number of Goroutines to use.
+
+```go
+    const grs = 2
+```
+
+wg is used to manage concurrency.
+
+```go
+    var wg sync.WaitGroup
+    wg.Add(grs)
+```
+
+Create two goroutines.
+
+```go
+    for i := 0; i < grs; i++ {
+        go func() {
+            for count := 0; count < 2; count++ {
+```
+
+Safely add one to the counter. Add the atomic functions that we have taken an address as the first parameter and that is being syncronized, no matter how many Goroutines they are. If we call one of these functions on the same location, they will get serialized. This the fastest way to serialization.
+
+We can run thist program all day long and still get 4 every time.
+
+```go
+                atomic.AddInt64(&counter, 1)
+```
+
+This call is now irrelevant because by the time AddInt64 function completes, counter is increment.
+
+```go
+                runtime.Gosched()
+            }
+
+        wg.Done()
+        }()
+    }
+```
+
+Wait for the Goroutines to finish.
+
+```go
+    wg.Wait()
+```
+
+Display the final value.
+
+```go
+    fmt.Println("Final Counter:", counter)
+}
+```
+
+```shell
+Final Counter: 4
+```
+
+##### Mutexes
+
+We don't always have the luxury of using 4-8 bytes of memory as shared data. This is where the mutex comes in. Mutex allows us to have an API like the WaitGroup (Add, Done and Wait) where any Goroutine can execute one at a time.
+
+```go
+package main
+
+import (
+    "fmt"
+    "sync"
+)
+
+var (
+```
+
+counter is a variable increment by all Goroutines.
+
+```go
+    counter int
+```
+
+mutex is used define a critical section of code. Picture mutex as a room where all Goroutines have to go through. However, only one Goroutine can go at a time. The scheduler will decide who can get in and which one is next. We cannot determine what the scheduler is gonna do. Hopefully, it is gonna be fair. Just because one Goroutine got to the door before another, it doesn't mean that Goroutine will get to the end first. Nothing here is predictable.
```

Comment:
```suggestion
mutex is used to define a critical section of code. Picture mutex as a room where all Goroutines have to go through. However, only one Goroutine can go at a time. The scheduler will decide who can get in and which one is next. We cannot determine what the scheduler is gonna do. Hopefully, it is gonna be fair. Just because one Goroutine got to the door before another, it doesn't mean that Goroutine will get to the end first. Nothing here is predictable.
```